test/telemetry/src/main/java/org/opensearch/test/telemetry/tracing/validators/NumberOfTraceIDsEqualToRequests.java

```diff
@@ -33,11 +40,18 @@ public NumberOfTraceIDsEqualToRequests() {}
      */
     @Override
     public List<MockSpanData> validate(List<MockSpanData> spans, int requests) {
-        Set<String> totalTraceIDs = spans.stream().map(MockSpanData::getTraceID).collect(Collectors.toSet());
+        Set<String> totalTraceIDs = spans.stream()
+            .filter(span -> isMatchingSpan(span))
+            .map(MockSpanData::getTraceID)
+            .collect(Collectors.toSet());
         List<MockSpanData> problematicSpans = new ArrayList<>();
         if (totalTraceIDs.size() != requests) {
             problematicSpans.addAll(spans);
         }
         return problematicSpans;
     }
+
+    private boolean isMatchingSpan(MockSpanData mockSpanData) {
+        return mockSpanData.getAttributes().get(FILTERING_ATTRIBUTE).equals(attributes.getAttributesMap().get(FILTERING_ATTRIBUTE));
```

Comment:
mockSpanData.getAttributes().get(FILTERING_ATTRIBUTE) will NPE if attribute is not there:

```suggestion
        if (attributes.isEmpty()) {
            return true;
        }
            
        return  Objects.equals(mockSpanData.getAttributes().get(FILTERING_ATTRIBUTE), attributes.getAttributesMap().get(FILTERING_ATTRIBUTE));
```